Refactor KMSKeyHandle controller

gemmahou opened the pull request:
### Change description

<!--
Describe what this pull request does.

* If your pull request is to address an open issue, indicate it by specifying the
issue number: 

For example: "Fixes #858"
-->

- `keyHandleID` can be either specified by user or generated by the server. Refactor the code to follow the pattern when resourceID(keyHandleID) is server generated.
- Remove redundant code
- Error out when trying to modify the resource, since Update is not supported. 
- update KRM status even there's no update.
- update test yaml and record real&mockGCP logs
- update doc with prerequisites
- be able to delete the KCC resource(no-op to GCP)

#### Special notes for your reviewer:

#### Does this PR add something which needs to be 'release noted'?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
## Reconciliation Improvements
*   [`KMSKeyHandle`](https://cloud.google.com/config-connector/docs/reference/resource-docs/kms/kmskeyhandle)

    *   Added status.observedState.kmsKey with the full path of the autokey.
    *   Update to an autokey is not allowed by the API. An error will be logged when attempting to modify the resource.
    *   Delete an autokey is not supported by the API. Delete the corresponding KCC resource is supported(no-op to GCP).
```

- [ ] Reviewer reviewed release note.

#### Additional documentation e.g., references, usage docs, etc.:

<!--
This section can be blank if this pull request does not require any additional documentation.

When adding links which point to resources within git repositories, like
usage documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
Added a new "Prerequisites" section to the [resource documentation](https://cloud.google.com/config-connector/docs/reference/resource-docs/kms/kmskeyhandle) with required setup steps.
```

#### Intended Milestone

Please indicate the intended milestone. 
- [ ] Reviewer tagged PR with the actual milestone.

### Tests you have done

<!--

Make sure you have run "make ready-pr" to run required tests and ensure this PR is ready to review. 

Also if possible, share a bit more on the tests you have done. 

For example if you have updated the pubsubtopic sample, you can share the test logs from running the test case locally.

go test -v -tags=integration ./config/tests/samples/create -test.run TestAll -run-tests pubsubtopic

-->

- [ ] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

maqiuyujoyce commented on pkg/controller/direct/kms/keyhandle/keyhandle_controller.go:
```diff
@@ -168,9 +176,34 @@ func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperati
+		status.ExternalRef = &externalRef
+		return updateOp.UpdateStatus(ctx, status, nil)
+	} else {
+		return fmt.Errorf("update operation not supported for resource %v %v",
```
Not a blocker, but would be great if we make it explicit in the error message about what field got updated unexpectedly.

maqiuyujoyce commented on pkg/controller/direct/kms/keyhandle/keyhandle_mapper.go:
```diff
@@ -17,59 +17,26 @@ package keyhandle
 	out.ResourceTypeSelector = *in.ResourceTypeSelector
 	return out
 }
-func KeyHandle_FromProto(mapCtx *direct.MapContext, in *pb.KeyHandle) *krm.KeyHandle {
```
Looks like krm.KeyHandle is still a type? Can we clean it up?

gemmahou replied:
krm.KeyHandle is generated, I'm afraid even we remove it, when we run the generator on KMS resource it will get generated again. I feel it's fine to keep it in types_generated.go.

maqiuyujoyce commented on pkg/test/resourcefixture/testdata/basic/kms/v1beta1/kmskeyhandle/kmskeyhandle-with-id/_http.log:
```diff
@@ -16,6 +16,13 @@ X-Xss-Protection: 0
 {
   "error": {
     "code": 404,
+    "errors": [
```
Could you make the mock http log match the real one here?

maqiuyujoyce commented on pkg/test/resourcefixture/testdata/basic/kms/v1beta1/kmskeyhandle/kmskeyhandle-with-id/_http.log:
```diff
@@ -69,10 +76,13 @@ X-Xss-Protection: 0
 
 {
   "done": true,
+  "metadata": {
```
ditto

gemmahou replied:
I can fix this but I think it's unnecessary. This operation metadata comes from above POST request, our mock operation default to populate metadata in logs: 
As long as it's operation log, and unrelated to the API behavior, I feel it's probably fine.

I could fix this but I think it's unnecessary. This operation metadata comes from above POST request, mock operation already defaults to populating metadata in the logs(https://github.com/GoogleCloudPlatform/k8s-config-connector/blob/f51a53a3054479be5095790fab6381dc2f4848db/mockgcp/common/operations/operations.go#L83). Since this pertains to operation logs and doesn't affect API behavior, I think it's probably fine to leave it.

maqiuyujoyce replied:
It's okay to be like this for now, but the operation is also part of the API behavior and can impact how things work. I'd try to match all the logs for `https://cloudkms.googleapis.com` that is related to KMS key handle.

gemmahou replied:
Ah I forgot to mention that I removed this in ["address comments" commit](https://github.com/GoogleCloudPlatform/k8s-config-connector/pull/4339/commits/6e0b37b268b1f8d59839ee7c37fa5c6129ddf428). Need to introduce a new option to remove metadata in GET operation, as the original function just copies metadata from the previous response. Let me know if that makes sense.

maqiuyujoyce commented on pkg/test/resourcefixture/testdata/basic/kms/v1beta1/kmskeyhandle/kmskeyhandle-with-id/_http.log:
```diff
@@ -69,10 +76,13 @@ X-Xss-Protection: 0
   "name": "projects/${projectId}/locations/us-central1/operations/${operationID}",
   "response": {
     "@type": "type.googleapis.com/google.cloud.kms.v1.KeyHandle",
-    "kmsKey": "projects/yuhou-key-project/locations/us-central1/keyRings/autokey/cryptoKeys/${projectNumber}-compute-disk-ac103725a174885c",
```
I'm curious why is your project not converted into `${projectId}` here?

gemmahou replied:
This is related to the manual setup I added. There are two projects: key_project is the one that saves all auto keys, and resource_project is the one that uses the auto key: https://cloud.google.com/kms/docs/enable-autokey#set-up-key-project

maqiuyujoyce commented on scripts/generate-google3-docs/resource-reference/templates/kms_kmskeyhandle.tmpl:
```diff
@@ -49,6 +49,14 @@
+Before you can use this resource, you must follow the guide https://cloud.google.com/kms/docs/enable-autokey
+to enable and configure Cloud KMS Autokey on a resource folder.
+
+This includes: setup resource folder, setup key project, enable Cloud KMS Autokey on the resource folder,
```
Nit: Format the tmpl for .md file so that it wraps at the same width.

gemmahou commented on apis/kms/v1beta1/keyhandle_identity.go:
```diff
@@ -65,15 +69,8 @@ func NewKMSKeyHandleIdentity(ctx context.Context, reader client.Reader, obj *KMS
-	desiredHandleID := valueOf(obj.Spec.ResourceID)
-	// set the desired keyhandle id in the identity
-	id.id = desiredHandleID
+	resourceID := valueOf(obj.Spec.ResourceID)
```
I created a GH issue: https://github.com/GoogleCloudPlatform/k8s-config-connector/issues/4390

gemmahou commented:
/hold
Let's put this PR on hold to allow for further discussion regarding the necessity of the refactor.
The reasons I think we need this PR:
1. Cleaned up the code, removed special handling, redundant code and comments.
2. Adds required prerequisites to the g3 doc.
3. Fixed the Delete function in the direct controller so we can delete the KCC resource(no-op to GCP)

The reason that the refactor might be risky:
1. The resource is currently in Beta and considered functional.

gemmahou commented:
I think this PR is required if we want to support KMSKeyHandle(autokey). Currently KMSKeyHandleis not referenced by any resources, so I feel it's safe to make the update to the controller. This PR did not change existing logic and should not break customers.

maqiuyujoyce commented on pkg/controller/direct/kms/keyhandle/keyhandle_controller.go:
```diff
@@ -168,9 +176,34 @@ func (a *Adapter) Create(ctx context.Context, createOp *directbase.CreateOperati
-	log.V(2).Info("update operation not supported for KeyHandle resource")
-	return nil
+	log := klog.FromContext(ctx)
+	log.V(2).Info("updating Logging Link", "name", a.id)
```
Could you fix the log?

maqiuyujoyce reviewed (commented):
/approve
/hold

Overall looks good to me with some nit to fix.

gemmahou commented on pkg/test/resourcefixture/testdata/basic/kms/v1beta1/kmskeyhandle/kmskeyhandle-with-id/_generated_object_kmskeyhandle-with-id.golden.yaml:
```diff
@@ -0,0 +1,28 @@
+  - cnrm.cloud.google.com/deletion-defender
+  generation: 1
+  labels:
+    cnrm-test: "true"
```
With #5024 we may not require to add invalid labels test manually.

maqiuyujoyce reviewed (commented):
/lgtm
/approve
